File: lumi/gui/static/employee_tasks.js
/* Bounded SONN task controls. Private answers, journals and credentials stay on the host. */
window.LumiEmployeeTasks = class LumiEmployeeTasks {
    bindEmployeeTaskPanel() {
        document.getElementById('employee-task-button')?.addEventListener('click', () => this.openEmployeeTaskPanel());
    }

    openEmployeeTaskPanel() {
        document.getElementById('employee-task-dialog')?.remove();
        const overlay = document.createElement('div');
        overlay.id = 'employee-task-dialog';
        overlay.className = 'dialog-overlay';
        overlay.setAttribute('role', 'dialog');
        overlay.setAttribute('aria-modal', 'true');
        overlay.setAttribute('aria-label', 'Employee task');
        overlay.innerHTML = `<div class="dialog">
            <div class="dialog-header"><h2>Employee task</h2><button type="button" class="dialog-btn deny" data-close aria-label="Close employee task">×</button></div>
            <div class="dialog-body" data-task-content></div></div>`;
        // Styles go through element.style: the page's CSP refuses style="".
        Object.assign(overlay.querySelector('.dialog').style, {width: 'min(560px,94vw)', maxHeight: '90vh', overflow: 'auto'});
        const close = () => {
            overlay.remove();
            this._employeeTaskRequest = null;
            document.getElementById('employee-task-button')?.focus();
        };
        overlay.querySelector('[data-close]').addEventListener('click', close);
        overlay.addEventListener('keydown', event => {
            if (event.key === 'Escape') close();
            if (event.key === 'Tab') {
                const controls = [...overlay.querySelectorAll('button,input,select,textarea')].filter(node => !node.disabled && !node.hidden);
                const first = controls[0], last = controls[controls.length - 1];
                if (event.shiftKey && document.activeElement === first) { event.preventDefault(); last?.focus(); }
                else if (!event.shiftKey && document.activeElement === last) { event.preventDefault(); first?.focus(); }
            }
        });
        document.body.appendChild(overlay);
        overlay.querySelector('[data-close]').focus();
        if (!this.currentSessionId || this.currentBackendName !== 'sonn') {
            overlay.querySelector('[data-task-content]').textContent = 'Open a saved SONN conversation to manage its bounded employee task.';
            return;
        }
        this.requestEmployeeTask('view');
    }

    requestEmployeeTask(action, extra = {}) {
        const content = document.querySelector('#employee-task-dialog [data-task-content]');
        if (!content || this._employeeTaskRequest) return;
        const request = {command: 'employee_task', action, project: this.currentCwd,
            session_id: this.currentSessionId, request_id: crypto.randomUUID(), ...extra};
        this._employeeTaskRequest = request;
        content.querySelectorAll('button,input,select,textarea').forEach(node => {
            node.dataset.taskWasDisabled = String(node.disabled); node.disabled = true;
        });
        let status = content.querySelector('[data-status]');
        if (!status) { status = document.createElement('p'); status.dataset.status = ''; content.appendChild(status); }
        status.setAttribute('role', 'status');
        status.textContent = action === 'advice' ? 'Consulting the approved frontier model…' : 'Loading task state…';
        this.send(request);
    }

    receiveEmployeeTaskState(event) {
        const pending = this._employeeTaskRequest;
        if (!pending || event.request_id !== pending.request_id) return;
        this._employeeTaskRequest = null;
        const content = document.querySelector('#employee-task-dialog [data-task-content]');
        if (!content) return;
        if (event.session_id !== this.currentSessionId || this._normalizeProjectPath(event.project) !== this._normalizeProjectPath(this.currentCwd)) {
            content.textContent = 'The selected conversation changed. Close and reopen its task panel.';
            return;
        }
        if (event.error) {
            content.querySelectorAll('button,input,select,textarea').forEach(node => { node.disabled = node.dataset.taskWasDisabled === 'true'; });
            content.querySelector('[data-status]').textContent = event.error;
            if (!content.querySelector('button')) this.employeeTaskButton(content, 'Refresh task', () => this.requestEmployeeTask('view'));
            return;
        }
        content.replaceChildren();
        document.querySelector('#employee-task-dialog [data-close]')?.focus();
        const paragraph = text => { const p = document.createElement('p'); p.textContent = text; content.appendChild(p); };
        if (!event.available) { paragraph(event.message || 'Task controls are unavailable.'); return; }
        const task = event.task;
        if (!task) {
            if (event.used) {
                paragraph('This task is closed and the conversation uses ordinary per-request limits. Start a fresh conversation for another bounded task.');
                return;
            }
            paragraph('Set one total allowance for the next task. All execution, compression and advice calls share it: at most eight calls, including two consultations, within 15 minutes.');
            const form = document.createElement('form');
            form.innerHTML = `<label>Total task allowance (USD)<input class="settings-input" name="allowance" type="number" min="0.01" max="20" step="0.01" value="2" required></label>
                <label>Task complexity<select class="settings-select" name="complexity"><option value="unknown">Not specified</option><option value="routine">Routine</option><option value="complex">Complex with several dependencies</option></select></label>
                <p>Your saved project model policy applies. The server also enforces your project's request ceiling.</p>
                <button type="submit" class="dialog-btn allow">Start bounded task</button>`;
            form.querySelectorAll('label').forEach(label => { label.style.display = 'grid'; label.style.gap = '8px'; label.style.marginBottom = '16px'; });
            form.addEventListener('submit', event => {
                event.preventDefault();
                const amount = Number(form.elements.allowance.value);
                if (!Number.isFinite(amount) || amount <= 0 || amount > 20) return;
                const choice = form.elements.complexity.value;
                const descriptor = choice === 'unknown' ? {} : choice === 'routine'
                    ? {task_breadth: 1, dependency_breadth: 0, contract_ambiguity: 0}
                    : {task_breadth: 6, dependency_breadth: 6, contract_ambiguity: 0.5};
                this.requestEmployeeTask('start', {ceiling_microusd: Math.round(amount * 1000000), descriptor});
            });
            content.appendChild(form);
            return;
        }
        const quota = task.quota || {};
        paragraph(`Status: ${task.status}${task.deadline_passed ? ' · deadline passed' : ''}`);
        const identity = document.createElement('details');
        const summary = document.createElement('summary'); summary.textContent = 'Task identity';
        const identifier = document.createElement('code'); identifier.textContent = task.id;
        identity.append(summary, identifier); content.appendChild(identity);
        paragraph(`Charged $${((quota.charged_microusd || 0) / 1000000).toFixed(4)} of $${(task.ceiling_microusd / 1000000).toFixed(2)}. Reserved allowance: $${((quota.allocated_microusd || 0) / 1000000).toFixed(4)}.`);
        paragraph(`Calls used or reserved: ${quota.claimed_calls || 0}/${task.call_limit}. Consultations: ${quota.claimed_advisory || 0}/${task.advisory_limit}.`);
        paragraph(`Deadline: ${new Date(task.deadline * 1000).toLocaleString()}`);
        if (event.pending || task.unresolved_attempts || task.unresolved_preflights) paragraph('An original request needs reconciliation. Recovery checks its saved result and does not repeat a model call.');
        if (event.advice_status === 'awaiting') paragraph('Recorded advice is ready for the next execution. Compression will not consume it.');
        this.employeeTaskButton(content, 'Refresh task', () => this.requestEmployeeTask('view'));
        this.employeeTaskButton(content, 'Recover original requests', () => this.requestEmployeeTask('recover'));
        this.employeeTaskButton(content, 'Inspect assignments', () => this.requestEmployeeTask('graph'));
        if (Object.hasOwn(event, 'graph')) this.renderEmployeeAssignments(content, event.graph);
        if (task.status === 'active' && !task.deadline_passed) {
            this.employeeTaskButton(content, 'Cancel task', () => this.requestEmployeeTask('cancel'));
            const form = document.createElement('form');
            form.innerHTML = `<label>Consultation purpose<select class="settings-select" name="purpose"><option value="advise">Execution advice</option><option value="coordinate">Employee coordination</option></select></label>
                <label>Question for the frontier advisor<textarea class="settings-input" name="question" rows="4" maxlength="8192" required></textarea></label>
                <p>Requires your saved frontier-advice consent. One consultation uses the task's allowance. Its answer is unverified advice for the next execution.</p>
                <button type="submit" class="dialog-btn allow">Consult frontier advisor</button>`;
            Object.assign(form.querySelector('select').style, {display: 'block', width: '100%', margin: '8px 0 16px'});
            Object.assign(form.elements.question.style, {display: 'block', width: '100%', boxSizing: 'border-box'});
            const canConsult = !event.pending && event.advice_status !== 'awaiting' && event.advice_status !== 'following' && (quota.claimed_advisory || 0) < task.advisory_limit;
            form.querySelector('button').disabled = !canConsult;
            form.addEventListener('submit', submit => { submit.preventDefault(); this.requestEmployeeTask('advice', {
                question: form.elements.question.value, consultation_mode: form.elements.purpose.value}); });
            content.appendChild(form);
        } else {
            this.employeeTaskButton(content, 'Return to ordinary requests', () => this.requestEmployeeTask('detach'));
        }
    }

    renderEmployeeAssignments(content, graph) {
        const section = document.createElement('section');
        const heading = document.createElement('h3'); heading.textContent = 'Assignments';
        section.appendChild(heading); content.appendChild(section);
        const line = (parent, text) => { const p = document.createElement('p'); p.textContent = text; parent.appendChild(p); };
        if (!graph) { line(section, 'This task has no specialist assignments.'); return; }
        line(section, `Graph: ${graph.status}. Parent task: ${graph.root_status}${graph.deadline_passed ? ' · deadline passed' : ''}.`);
        const list = document.createElement('ol'); section.appendChild(list);
        for (const node of graph.nodes) {
            const item = document.createElement('li'); item.style.overflowWrap = 'anywhere'; list.appendChild(item);
            const title = document.createElement('strong');
            title.textContent = `${node.id}${node.integration ? ' · coordinator integration' : ' · specialist'}`;
            item.appendChild(title);
            line(item, `Employee: ${node.employee_id}. Status: ${node.status}. Execution: ${node.epoch}.`);
            if (node.status === 'reported') line(item, 'Output reported. Independent verification is still required.');
            if (node.status === 'failed') line(item, 'Verification failed. Host recovery is required before another execution.');
            if (node.lease_expired) line(item, 'Execution lease expired. The host must confirm the original worker stopped before reassigning it.');
            if (node.unresolved_file_actions) line(item, `${node.unresolved_file_actions} original file action(s) need host reconciliation.`);
            if (node.blocked_by.length) line(item, `Waiting for: ${node.blocked_by.join(', ')}.`);
            if (node.prior_executions) line(item, `Prior executions retained: ${node.prior_executions}.`);
            const details = document.createElement('details');
            const summary = document.createElement('summary'); summary.textContent = 'Assignment contract';
            details.appendChild(summary);
            line(details, node.input_contract);
            line(details, `Expected outputs: ${node.outputs.join(', ')}.`);
            line(details, `Dependencies: ${node.dependencies.join(', ') || 'None'}.`);
            item.appendChild(details);
        }
    }

    employeeTaskButton(parent, label, action) {
        const button = document.createElement('button');
        button.className = 'dialog-btn deny';
        button.type = 'button'; button.textContent = label;
        button.style.margin = '8px 8px 16px 0';
        button.addEventListener('click', action); parent.appendChild(button);
    }
};

File: lumi/gui/employee_tasks.py
"""Native UI task controls with private durable configuration and captured identity."""
from __future__ import annotations

import json
from contextlib import closing
from pathlib import Path
import sqlite3
import time

from ..sonn import SonnBackend
from ..sonn_tasks import SonnTaskController, SonnTaskError
from .runtime import bind_sonn_conversation
from .sessions import _sessions_dir, is_valid_session_id


def journal_path(project_path: str, session_id: str) -> Path:
    if not project_path or not is_valid_session_id(session_id):
        raise SonnTaskError('Open a saved SONN conversation to manage a bounded task')
    return _sessions_dir(project_path) / (session_id + '.sonn-task.sqlite')


def restore_task(backend, project_path: str, session_id: str):
    """Restore before generation. Invalid saved setup must never fall back to unbounded calls."""
    if not isinstance(backend, SonnBackend):
        return None
    path = journal_path(project_path, session_id)
    if not path.exists():
        if backend.task_controller is not None:
            raise SonnTaskError('This backend still belongs to another task; reopen the intended conversation')
        return None
    try:
        with closing(sqlite3.connect(path.resolve().as_uri() + '?mode=ro', uri=True)) as connection:
            row = connection.execute('SELECT value FROM task WHERE id=1').fetchone()
        state = json.loads(row[0]) if row else {}
        if state.get('detached'):
            backend.task_controller = None
            return None
        config = state['configuration']
        bind_sonn_conversation(backend, project_path, session_id)
        controller = SonnTaskController(backend, path, ceiling_microusd=config['ceiling_microusd'],
                                        descriptor=config['descriptor'])
        backend.task_controller = controller
        return controller
    except (sqlite3.Error, OSError, ValueError, KeyError, TypeError) as exc:
        raise SonnTaskError('Saved employee task needs recovery before this conversation can generate again') from exc


def configure(backend, project_path: str, session_id: str, *, ceiling_microusd: int, descriptor: dict):
    if not isinstance(backend, SonnBackend):
        raise SonnTaskError('Choose the SONN backend for this conversation')
    path = journal_path(project_path, session_id)
    bind_sonn_conversation(backend, project_path, session_id)
    if path.exists():
        controller = restore_task(backend, project_path, session_id)
        if controller is None:
            raise SonnTaskError('This conversation already has a closed task; start a fresh conversation for another')
        controller = controller.revise_unstarted(ceiling_microusd=ceiling_microusd, descriptor=descriptor)
    else:
        controller = SonnTaskController(backend, path, ceiling_microusd=ceiling_microusd, descriptor=descriptor)
    backend.task_controller = controller  # Attach before networking; interruptions stay bounded.
    controller.start()
    return controller


def public_state(backend, project_path: str, session_id: str):
    if not isinstance(backend, SonnBackend):
        return {'available': False, 'message': 'Choose a saved SONN conversation to manage a bounded employee task.'}
    controller = restore_task(backend, project_path, session_id)
    if controller is None:
        return {'available': True, 'task': None, 'used': journal_path(project_path, session_id).exists()}
    local = controller.state()
    task = checked_view(controller)
    # Explicit projection; never send credentials, local advice text, grants,
    # private journal paths, or arbitrary server records to the browser.
    return {'available': True, 'task': {key: task.get(key) for key in (
        'id', 'status', 'deadline', 'deadline_passed', 'ceiling_microusd', 'call_limit', 'advisory_limit',
        'employee_id', 'unresolved_attempts', 'unresolved_preflights')} | {'quota': {
            key: (task.get('quota') or {}).get(key, 0) for key in
            ('charged_microusd', 'allocated_microusd', 'claimed_calls', 'claimed_advisory')}},
        'pending': bool(local.get('pending')), 'advice_status': (local.get('advice') or {}).get('status'),
        'configuration': local['configuration']}


def detach(controller):
    """Explicitly return to normal per-request work after closing the old root."""
    controller.recover()
    task = checked_view(controller)
    if (task['status'] not in ('cancelled', 'completed') and not task.get('deadline_passed')
            or task.get('unresolved_attempts') or task.get('unresolved_preflights')
            or task.get('quota', {}).get('allocated_microusd')):
        raise SonnTaskError('Close the task and reconcile every original request before returning to ordinary requests')
    controller._update(lambda state: {**state, 'detached': True})
    controller.backend.task_controller = None


def checked_view(controller):
    task = controller.view()
    quota = task.get('quota')
    fields = ('allocated_microusd', 'charged_microusd', 'claimed_calls', 'claimed_advisory')
    if (task.get('id') != controller.state()['root_id'] or not isinstance(quota, dict)
            or any(type(quota.get(key)) is not int or quota[key] < 0 for key in fields)
            or any(type(task.get(key)) is not int or task[key] < 0 for key in ('unresolved_attempts', 'unresolved_preflights'))):
        raise SonnTaskError('Task accounting is incomplete; retain the original task and reconcile before continuing')
    return task


def recover_all(controller):
    """Reconcile saved root attempts, including specialists, without model retries."""
    task = checked_view(controller)
    failures = []
    attempts = task.get('attempts')
    if not isinstance(attempts, list) or len(attempts) > 128:
        raise SonnTaskError('Root recovery requires its bounded original attempt list')
    for attempt in attempts:
        if attempt.get('status') in ('allocated', 'reserved', 'refusing', 'dispatching', 'completed'):
            try:
                controller._request('POST', '/' + task['id'] + '/recover', {'attempt_id': attempt['id']})
            except SonnTaskError as exc:
                failures.append(exc)
    try:
        controller.recover()
    except SonnTaskError as exc:
        failures.append(exc)
    if failures:
        raise SonnTaskError('Some original requests remain unresolved. Known receipts were reconciled; refresh task status before continuing.')


def public_graph(controller):
    """Expose assignment contracts and recovery state, never worker capabilities."""
    graph = controller.graph().get('graph')
    if graph is None:
        return None
    nodes = graph.get('nodes') if isinstance(graph, dict) else None
    if not isinstance(nodes, dict) or not 1 <= len(nodes) <= 8:
        raise SonnTaskError('Assignment graph is incomplete; refresh before continuing')
    result = []
    for key, node in nodes.items():
        if (not isinstance(node, dict) or node.get('id') != key
                or node.get('status') not in ('waiting', 'running', 'reported', 'failed', 'succeeded')
                or not isinstance(node.get('dependencies'), list) or len(node['dependencies']) > 7
                or any(dependency not in nodes for dependency in node['dependencies'])
                or not isinstance(node.get('outputs'), list) or len(node['outputs']) > 8):
            raise SonnTaskError('Assignment graph is incomplete; refresh before continuing')
        expired = node['status'] == 'running' and time.time() >= node.get('lease_until', 0)
        pending_tools = sum(action.get('status') != 'reported' for action in node.get('tool_actions', {}).values()
                            if action.get('identity', {}).get('epoch') == node.get('epoch'))
        result.append({field: node.get(field) for field in
            ('id', 'employee_id', 'input_contract', 'outputs', 'dependencies', 'integration', 'status', 'epoch')} | {
                'lease_expired': expired, 'unresolved_file_actions': pending_tools,
                'prior_executions': len(node.get('history', [])),
                'blocked_by': [dependency for dependency in node['dependencies'] if nodes[dependency]['status'] != 'succeeded']})
    return {'status': graph.get('status'), 'root_status': graph.get('root_status'),
            'deadline_passed': graph.get('deadline_passed'), 'nodes': result}


async def command(state, send, message):
    """Run through the chat queue so setup/advice cannot race ordinary execution."""
    import asyncio
    import os
    project = state.project.project_path
    record = state.project.current_session
    session_id = record.id if record else ''
    reply = {'event': 'employee_task_state', 'project': project, 'session_id': session_id,
             'request_id': str(message.get('request_id') or '')}
    try:
        if (not session_id or message.get('session_id') != session_id
                or os.path.normcase(os.path.abspath(str(message.get('project') or ''))) != os.path.normcase(os.path.abspath(project))):
            raise SonnTaskError('The selected conversation changed; reopen its employee task panel')
        if state.session is None:
            raise SonnTaskError('Open a saved SONN conversation before changing its task')
        backend = state.session.backend
        action = message.get('action', 'view')
        def operate():
            assignments = {}
            if action == 'start':
                configure(backend, project, session_id, ceiling_microusd=message.get('ceiling_microusd'),
                          descriptor=message.get('descriptor', {}))
            elif action != 'view':
                controller = restore_task(backend, project, session_id)
                if controller is None:
                    raise SonnTaskError('This conversation has no active bounded task')
                if action == 'recover':
                    recover_all(controller)
                elif action == 'graph':
                    assignments['graph'] = public_graph(controller)
                elif action == 'cancel':
                    controller.cancel()
                elif action == 'advice':
                    controller.ask_advice(message.get('question'), mode=message.get('consultation_mode', 'advise'))
                elif action == 'detach':
                    detach(controller)
                else:
                    raise SonnTaskError('Unknown employee task action')
            return public_state(backend, project, session_id) | assignments
        reply.update(await asyncio.to_thread(operate))
    except (SonnTaskError, ValueError, OSError, sqlite3.Error) as exc:
        reply.update(error=str(exc), available=False)
    await send(reply)
